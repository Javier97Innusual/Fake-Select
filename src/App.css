.App {
  display: flex;
  flex-flow: column;
  justify-content: center;
  align-items: center;
  gap: 2rem;
  padding: 2rem;

  form {
    display: flex;
    gap: .5rem;
  }

  .rectangle-main {
    width: 100px;
    height: 80px;

    &.rectangle-red {
      background: red;
    }

    &.rectangle-blue {
      background: blue;
    }

    &.rectangle-green {
      background: green;
    }
  }
}