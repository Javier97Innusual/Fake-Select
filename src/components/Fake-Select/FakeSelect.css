.select-main {
    display: flex;
    flex-flow: column;
    width: 400px;
    position: relative;

    .select-current {
        cursor: pointer;
        padding: 1.5rem 1rem;
        border: 2px solid black;
        font-size: 1.225rem;
        font-weight: bold;
        text-align: left;
    }

    .select-container {
        display: flex;
        flex-flow: column;
        border: 2px solid black;
        border-top: none;
    }

    &::after {
        content: '▼';
        position: absolute;
        right: 1rem;
        top: 1.725rem;
    }

    &.is-active {
        &::after {
            content: '✖';
        }
    }
}