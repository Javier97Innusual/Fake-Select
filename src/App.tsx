import './App.css';
import { FakeOption } from './components/Fake-Option/FakeOption';
import { FakeSelect } from './components/Fake-Select/FakeSelect';

function App() {
  return (
    <div className="App">
      <FakeSelect
        name="SelectName"
        disabled={true}
      >
        <FakeOption value="red">Rojo</FakeOption>
        <FakeOption value="blue">Azul</FakeOption>
        <FakeOption value="green">Verde</FakeOption>
      </FakeSelect>
    </div>
  );
}

export default App;
