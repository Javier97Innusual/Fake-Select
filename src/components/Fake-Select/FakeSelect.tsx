import { MouseEvent, useState } from 'react';
import classNames from 'classnames';
import { DEFAULT_VALUE } from './FakeSelect.constants';
import { FakeSelectType, OptionValueType } from './FakeSelect.type';
import './FakeSelect.css';

export function FakeSelect({ id, placeholder, onChange, disabled, children }: FakeSelectType) {
    const [isSelectActive, setSelectActive] = useState<boolean>(false);
    const [currentOption, setCurrentOption] = useState<OptionValueType>({
        value: '',
        name: placeholder || DEFAULT_VALUE
    });

    const handleOption = (event: MouseEvent) => {
        const target = event.target as HTMLElement;

        if (target.dataset.id === undefined || target.innerHTML === null || disabled) {
            return;
        }

        setCurrentOption({
            value: target.dataset.id,
            name: target.innerText
        });

        onChange(target.dataset.id);

        setSelectActive(!isSelectActive);
    }

    return (
        <div
            id={id}
            className={classNames('select-main', { 'is-active': !disabled && isSelectActive })}
        >
            <div
                className="select-current"
                onClick={() => setSelectActive(!isSelectActive)}
            >
                {currentOption.name}
            </div>
            {!disabled && isSelectActive && (
                <>
                    <div
                        className="select-container"
                        onClick={handleOption}
                    >
                        {children}
                    </div>
                    <div
                        className="select-background"
                        onClick={() => setSelectActive(!isSelectActive)}
                    ></div>
                </>
            )}
        </div>
    )
}