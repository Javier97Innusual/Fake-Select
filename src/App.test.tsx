import { render, fireEvent } from '@testing-library/react';
import App from './App';
import { DEFAULT_VALUE } from './components/Fake-Select/FakeSelect.constants';

describe('App component', () => {
  it('should render correctly and match snapshot', () => {
    const { asFragment } = render(
      <App />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should not display rectangle on initial render', () => {
    const { container } = render(<App />);

    expect(container.querySelector('.rectangle-main')).not.toBeInTheDocument();
  });

  it('should update state and displays rectangle on form submission', () => {
    const { getByText, container } = render(<App />);

    fireEvent.click(getByText(DEFAULT_VALUE));
    fireEvent.click(getByText('Rojo'));
    fireEvent.click(getByText('Submit'));

    expect(container.querySelector('.rectangle-main')).toHaveClass('rectangle-red');
  });

  it('should update state correctly when a different option is selected', () => {
    const { getByText, container } = render(<App />);

    fireEvent.click(getByText(DEFAULT_VALUE));
    fireEvent.click(getByText('Azul'));
    fireEvent.click(getByText('Submit'));

    expect(container.querySelector('.rectangle-main')).toHaveClass('rectangle-blue');
  });

  it('should not submit form if no option is selected', () => {
    const { getByText, container } = render(<App />);

    fireEvent.click(getByText('Submit'));

    expect(container.querySelector('.rectangle-main')).not.toBeInTheDocument();
  });
});