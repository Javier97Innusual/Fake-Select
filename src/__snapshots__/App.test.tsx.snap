// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render correctly and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <form
      id="test-form"
      method="get"
    >
      <div
        class="select-main"
        id="SelectName"
      >
        <div
          class="select-current"
        >
          Elije una opci√≥n
        </div>
      </div>
      <button
        form="test-form"
        type="submit"
        value="Submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
